Fix PDA seed path resolution for dot notation (account.field)

simplysabir opened the pull request:
## Description :
This PR fixes the panic that occurs when processing PDA seeds with dot notation. paths like "account.owner". The fix enhances the path resolution logic to properly parse and handle account field references.

## Changes:
- Enhanced OUTER loop in genAccountGettersSetters to handle dot notation
- Added proper parsing for "account.field" patterns
- Added detailed logging for debugging dot notation resolution failures
- Maintains backward compatibility with simple account name paths

Fixes: #18 

## Technical Note : 
For dot notation like "account.owner", we resolve the account ("account") and defer field access to runtime PDA derivation. The actual field access happens when the PDA is derived using the account's current state, not during code generation. This matches Anchor's runtime behavior.

## Test : 
- To reproduce use this idl on current version of `solana-anchor-go` and then with this updated one
```json
{
  "address": "TestProgram1111111111111111111111111111111",
  "metadata": {
    "name": "test_program",
    "version": "0.1.0",
    "spec": "0.1.0"
  },
  "instructions": [
    {
      "name": "test_instruction",
      "accounts": [
        {
          "name": "user",
          "writable": true,
          "signer": true
        },
        {
          "name": "position",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [112, 111, 115, 105, 116, 105, 111, 110]
              },
              {
                "kind": "account",
                "path": "position.owner",
                "account": "Position"
              }
            ]
          }
        }
      ],
      "args": []
    }
  ],
  "accounts": [
    {
      "name": "Position",
      "discriminator": [1, 2, 3, 4, 5, 6, 7, 8]
    }
  ],
  "types": [
    {
      "name": "Position",
      "type": {
        "kind": "struct",
        "fields": [
          {
            "name": "owner",
            "type": "pubkey"
          }
        ]
      }
    }
  ]
}

```

simplysabir committed:
feat: fixed account resolution

simplysabir committed:
fix: added robust fallbacks and comments

simplysabir commented on main.go:
```diff
@@ -1583,6 +1583,24 @@ func genAccountGettersSetters(
+							}
+						}
+						// If we can't resolve the dot notation, fall through to panic
+					}
```
gotcha fixed that in the new commit

simplysabir commented on main.go:
```diff
@@ -1583,6 +1583,24 @@ func genAccountGettersSetters(
+								if acc.IdlAccount.Name == accountName {
+									// Use the account name, ignoring the field for now
+									// The field access will be handled in the generated code
+									seedRefs[i] = ToLowerCamel(acc.IdlAccount.Name)
```
explained this in the new commit

dehypnosis commented:
so how “ Note: Field access is deferred to runtime. When the PDA is derived, …” is achieved? is it just left to user?

simplysabir commented:
> so how “ Note: Field access is deferred to runtime. When the PDA is derived, …” is achieved? is it just left to user?

my bad I just managed to extract out the position, thanks for pointing it out. 

now implemented full support for dot-notation PDA seeds (account.field). the generator now produces helper functions that accept the deserialized account at runtime, automatically handling the necessary field access and serialization.

also hardened the IDL parser to be backwards-compatible, fixing two critical panics. It now correctly handles legacy formats, such as string-based type definitions and missing metadata objects, to improve stability.

simplysabir committed:
fix: support dot-notation PDA seeds and improve legacy IDL parsing

simplysabir commented:
> > so how “ Note: Field access is deferred to runtime. When the PDA is derived, …” is achieved? is it just left to user?
> 
> my bad I just managed to extract out the position, thanks for pointing it out.
> 
> now implemented full support for dot-notation PDA seeds (account.field). the generator now produces helper functions that accept the deserialized account at runtime, automatically handling the necessary field access and serialization.
> 
> also hardened the IDL parser to be backwards-compatible, fixing two critical panics. It now correctly handles legacy formats, such as string-based type definitions and missing metadata objects, to improve stability.

any updates @dehypnosis ?

dehypnosis commented:
@simplysabir so sorry for the delayed. I was so busy doing the business. plz lemme take a look asap

simplysabir commented:
> @simplysabir so sorry for the delayed. I was so busy doing the business. plz lemme take a look asap

yeah no worries!

simplysabir commented:
> @simplysabir so sorry for the delayed. I was so busy doing the business. plz lemme take a look asap

hey @dehypnosis any updates?

Tougee commented:
If `kind` is `arg`, `path` could also involve dot notation. Would you be open to supporting that too?

```json
{
          "name": "remote_token_messenger",
          "writable": true,
          "pda": {
            "seeds": [
              {
                "kind": "const",
                "value": [
                  114,
                  101
                ]
              },
              {
                "kind": "arg",
                "path": "params.domain"
              }
            ]
          }
        }
```

simplysabir commented:
> If `kind` is `arg`, `path` could also involve dot notation. Would you be open to supporting that too?
> 
> ```json
> {
>           "name": "remote_token_messenger",
>           "writable": true,
>           "pda": {
>             "seeds": [
>               {
>                 "kind": "const",
>                 "value": [
>                   114,
>                   101
>                 ]
>               },
>               {
>                 "kind": "arg",
>                 "path": "params.domain"
>               }
>             ]
>           }
>         }
> ```

okay lemme work on that

hooNode commented:
@simplysabir First of all, thank you for submitting such a thoughtful PR. We’ve been quite tied up on the business side, so I truly appreciate your patience while we took the time to review this.

I went through your PR thoroughly. The way you improved the structure for managing PDA seeds is much more efficient and readable—I found this particularly impressive. That said, I did identify a couple of potential issues, and I’d like to get your thoughts on them.

**1. Matching instruction args and "arg" seeds by name vs. index**

I noticed you’re validating `"arg"`-type seeds by checking `instruction.Args[i].Name == seedDef.Path`. While this is reasonable in many cases, in reality, with Anchor-generated IDLs, the instruction arg’s name and the seed’s name may not always match.
To be more robust, rather than relying on name equality, I think it’s better to match `"arg"`-type seeds and instruction args by index. For example:

```json
{
  "seeds": [
    { "kind": "arg", "path": "batch_id" }
  ],
  "args": [
    { "name": "_batch_id", "type": "u64" }
  ]
}
```

As you can see, the names don’t always match, but the order/index of `ins.args[i]` and `seedArgs[i]` will correspond.
So, I believe it would be more accurate to pair them by index instead of name for validation.

**2. Handling "account" type seeds and the optionality of `account`:**

For seeds of kind `"account"`, you’re using `accountTypeName := ToCamel(seedDef.Account)` to retrieve the account’s name. However, according to the [Anchor IDL spec](https://github.com/solana-foundation/anchor/blob/master/idl/spec/src/lib.rs):

```rust
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct IdlSeedAccount {
    pub path: String,
    #[serde(skip_serializing_if = "is_default")]
    pub account: Option<String>,
}
```

The `account` field here is optional, which means there can be edge cases where it is not set or is an empty string. For instance:

```json
"pda": {
  "seeds": [
    {
      "kind": "account",
      "path": "source_receipt_token_account.owner"
    }
  ]
}
```

In such cases, instruction parsing may not work as expected.
How do you think we should handle these scenarios where the `account` field is missing or empty?

Finally, I noticed that when running the `make` command, there’s an error during parsing with the example IDL.
Specifically, in the following code:

```go
Id(paramName+fieldName+"Bytes").Op(",").Err().Op(":=").Qual(PkgDfuseBinary, "MarshalBorsh").Call(Id(paramName).Dot(ToCamel(fieldName))),
```

It looks like the issue is caused by reusing the already assigned `err` variable, which results in a redeclaration error.
You should declare `Id(paramName+fieldName+"Bytes")` first and use `=` instead of `:=` for assignment.

---

Overall, I really like the direction and concept of your code, but I’d love to discuss these issues in more depth.
Is there anything else you’d like to discuss or suggest regarding these points?

Again, thank you for your patience, and I hope we can continue to collaborate more actively going forward.

simplysabir commented:
> @simplysabir First of all, thank you for submitting such a thoughtful PR. We’ve been quite tied up on the business side, so I truly appreciate your patience while we took the time to review this.
> 
> I went through your PR thoroughly. The way you improved the structure for managing PDA seeds is much more efficient and readable—I found this particularly impressive. That said, I did identify a couple of potential issues, and I’d like to get your thoughts on them.
> 
> **1. Matching instruction args and "arg" seeds by name vs. index**
> 
> I noticed you’re validating `"arg"`-type seeds by checking `instruction.Args[i].Name == seedDef.Path`. While this is reasonable in many cases, in reality, with Anchor-generated IDLs, the instruction arg’s name and the seed’s name may not always match. To be more robust, rather than relying on name equality, I think it’s better to match `"arg"`-type seeds and instruction args by index. For example:
> 
> ```json
> {
>   "seeds": [
>     { "kind": "arg", "path": "batch_id" }
>   ],
>   "args": [
>     { "name": "_batch_id", "type": "u64" }
>   ]
> }
> ```
> 
> As you can see, the names don’t always match, but the order/index of `ins.args[i]` and `seedArgs[i]` will correspond. So, I believe it would be more accurate to pair them by index instead of name for validation.
> 
> **2. Handling "account" type seeds and the optionality of `account`:**
> 
> For seeds of kind `"account"`, you’re using `accountTypeName := ToCamel(seedDef.Account)` to retrieve the account’s name. However, according to the [Anchor IDL spec](https://github.com/solana-foundation/anchor/blob/master/idl/spec/src/lib.rs?rgh-link-date=2025-08-04T04%3A58%3A11Z):
> 
> ```rust
> #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
> pub struct IdlSeedAccount {
>     pub path: String,
>     #[serde(skip_serializing_if = "is_default")]
>     pub account: Option<String>,
> }
> ```
> 
> The `account` field here is optional, which means there can be edge cases where it is not set or is an empty string. For instance:
> 
> ```json
> "pda": {
>   "seeds": [
>     {
>       "kind": "account",
>       "path": "source_receipt_token_account.owner"
>     }
>   ]
> }
> ```
> 
> In such cases, instruction parsing may not work as expected. How do you think we should handle these scenarios where the `account` field is missing or empty?
> 
> Finally, I noticed that when running the `make` command, there’s an error during parsing with the example IDL. Specifically, in the following code:
> 
> ```go
> Id(paramName+fieldName+"Bytes").Op(",").Err().Op(":=").Qual(PkgDfuseBinary, "MarshalBorsh").Call(Id(paramName).Dot(ToCamel(fieldName))),
> ```
> 
> It looks like the issue is caused by reusing the already assigned `err` variable, which results in a redeclaration error. You should declare `Id(paramName+fieldName+"Bytes")` first and use `=` instead of `:=` for assignment.
> 
> Overall, I really like the direction and concept of your code, but I’d love to discuss these issues in more depth. Is there anything else you’d like to discuss or suggest regarding these points?
> 
> Again, thank you for your patience, and I hope we can continue to collaborate more actively going forward.

hey agree with most of the points, been running busy a bit, by weekend I'll get to this and add the supported functionality, but yeah I agree with your observations, will push it shortly, thanks!

simplysabir committed:
fix: bunch of fixes

simplysabir commented:
> > @simplysabir First of all, thank you for submitting such a thoughtful PR. We’ve been quite tied up on the business side, so I truly appreciate your patience while we took the time to review this.
> > I went through your PR thoroughly. The way you improved the structure for managing PDA seeds is much more efficient and readable—I found this particularly impressive. That said, I did identify a couple of potential issues, and I’d like to get your thoughts on them.
> > **1. Matching instruction args and "arg" seeds by name vs. index**
> > I noticed you’re validating `"arg"`-type seeds by checking `instruction.Args[i].Name == seedDef.Path`. While this is reasonable in many cases, in reality, with Anchor-generated IDLs, the instruction arg’s name and the seed’s name may not always match. To be more robust, rather than relying on name equality, I think it’s better to match `"arg"`-type seeds and instruction args by index. For example:
> > ```json
> > {
> >   "seeds": [
> >     { "kind": "arg", "path": "batch_id" }
> >   ],
> >   "args": [
> >     { "name": "_batch_id", "type": "u64" }
> >   ]
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As you can see, the names don’t always match, but the order/index of `ins.args[i]` and `seedArgs[i]` will correspond. So, I believe it would be more accurate to pair them by index instead of name for validation.
> > **2. Handling "account" type seeds and the optionality of `account`:**
> > For seeds of kind `"account"`, you’re using `accountTypeName := ToCamel(seedDef.Account)` to retrieve the account’s name. However, according to the [Anchor IDL spec](https://github.com/solana-foundation/anchor/blob/master/idl/spec/src/lib.rs?rgh-link-date=2025-08-05T08%3A50%3A00Z):
> > ```rust
> > #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
> > pub struct IdlSeedAccount {
> >     pub path: String,
> >     #[serde(skip_serializing_if = "is_default")]
> >     pub account: Option<String>,
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `account` field here is optional, which means there can be edge cases where it is not set or is an empty string. For instance:
> > ```json
> > "pda": {
> >   "seeds": [
> >     {
> >       "kind": "account",
> >       "path": "source_receipt_token_account.owner"
> >     }
> >   ]
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In such cases, instruction parsing may not work as expected. How do you think we should handle these scenarios where the `account` field is missing or empty?
> > Finally, I noticed that when running the `make` command, there’s an error during parsing with the example IDL. Specifically, in the following code:
> > ```go
> > Id(paramName+fieldName+"Bytes").Op(",").Err().Op(":=").Qual(PkgDfuseBinary, "MarshalBorsh").Call(Id(paramName).Dot(ToCamel(fieldName))),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It looks like the issue is caused by reusing the already assigned `err` variable, which results in a redeclaration error. You should declare `Id(paramName+fieldName+"Bytes")` first and use `=` instead of `:=` for assignment.
> > Overall, I really like the direction and concept of your code, but I’d love to discuss these issues in more depth. Is there anything else you’d like to discuss or suggest regarding these points?
> > Again, thank you for your patience, and I hope we can continue to collaborate more actively going forward.
> 
> hey agree with most of the points, been running busy a bit, by weekend I'll get to this and add the supported functionality, but yeah I agree with your observations, will push it shortly, thanks!

@hooNode finally got the time to get back to this, so far I've pushed the fixes for everything we discussed. here's the list anyway

-  Fixed redeclaration error
- Enhanced arg seed matching (Index-based)
- Added arg dot notation support 
- Added detection and handling for "path": "params.domain" patterns in arg seeds
-  Fixed optional account field handling
-  Added null checks and fallback logic when seedDef.Account is empty

dehypnosis commented:
Dear, Thank you for your contribution!

Your commits has been merged from https://github.com/fragmetric-labs/solana-anchor-go/pull/21 with a few additional changes:
- Updated the account seed related logic so that any account seed definition now requires a pubkey type, as Anchor always generates an account-kind seed definition if and only if the type of a nested account field is pubkey.

You can use this update with `github.com/fragmetric-labs/solana-anchor-go v1.2.0`.
